mlflow/tracing/export/mlflow.py

```diff
@@ -70,24 +76,159 @@ def export(self, root_spans: Sequence[ReadableSpan]):
                 self._display_handler.display_traces([trace])
 
-            # Log the trace to MLflow
             self._log_trace(trace)
 
     def _log_trace(self, trace: Trace):
+        """Log the trace to MLflow backend."""
+        upload_tag_task = Task(
+            handler=self._client._upload_trace_spans_as_tag,
+            args=(trace.info, trace.data),
+            error_msg="Failed to log trace spans as tag to MLflow backend.",
+        )
+
+        upload_trace_data_task = Task(
+            handler=self._client._upload_trace_data,
+            args=(trace.info, trace.data),
+            error_msg="Failed to log trace to MLflow backend.",
+        )
+
+        upload_ended_trace_info_task = Task(
+            handler=self._client._upload_ended_trace_info,
+            args=(trace.info,),
+            error_msg="Failed to log trace to MLflow backend.",
+        )
+
+        if MLFLOW_ENABLE_ASYNC_LOGGING.get():
+            self._async_queue.put(upload_tag_task)
+            self._async_queue.put(upload_trace_data_task)
+            self._async_queue.put(upload_ended_trace_info_task)
+        else:
+            upload_tag_task.handle()
+            upload_trace_data_task.handle()
+            upload_ended_trace_info_task.handle()
+
+
+class Task:
+    """A dataclass to represent a task to be processed by the async trace export queue."""
+
+    def __init__(self, handler: Callable, args: Sequence, error_msg: str = ""):
+        self._handler = handler
+        self._args = args
+        self._error_msg = error_msg
+
+    def handle(self):
+        """Handle the task. This method must not raise any exception."""
         try:
-            self._client._upload_trace_spans_as_tag(trace.info, trace.data)
+            self._handler(*self._args)
         except Exception as e:
-            _logger.debug(f"Failed to log trace spans as tag to MLflow backend: {e}", exc_info=True)
-
-        # TODO: Make this async
-        # The trace is already updated in processor.on_end method
-        # so we just log to backend store here
-        try:
-            self._client._upload_trace_data(trace.info, trace.data)
-            self._client._upload_ended_trace_info(trace.info)
-        except Exception as e:
-            # avoid silent failures
             _logger.warning(
-                f"Failed to log trace to MLflow backend: {e}",
+                f"{self._error_msg} Error: {e}. For full traceback, set logging level to debug.",
                 exc_info=_logger.isEnabledFor(logging.DEBUG),
             )
+
+
+class AsyncTraceExportQueue:
+    """
+    This is a queue based run data processor that queue incoming data and process it using a single
+    worker thread. This class is used to process traces saving in async fashion.
+    """
+
+    def __init__(self, client) -> None:
+        self._queue: Queue[Task] = Queue()
+        self._client = client
+        self._lock = threading.RLock()
+
```

Comment:
Should we initialize the instance with self._trace_logging_thread and self._trace_logging_worker_threadpool to None in case of integrations with this handler preventing attribute errors when accessing if those attributes aren't defined yet?

Reply:
Hmm `None` will cause NPE so doesn't prevent errors for typical use case accessing those fields... Currently we only access them in `flush()`, which checks if the queue is active or not, so I think it's safe to assume they exist.